### PR TITLE
Make the cache work with rust-analyzer

### LINKED ISSUE #22: Cache strategy hits assertion in Rust-Analyzer
After my weekly dependency updates this morning, I'm having async-graphql macros failing in my crates with this assertion:

```rust
thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
  left: `"/Users/djc/src/bolt/rust/domains/epoxide"`,
 right: `"/Users/djc/src/bolt/rust/domains/store"`: CARGO_MANIFEST_DIR must not change within one compiler process', /Users/djc/.cargo/registry/src/github.com-1ecc6299db9ec823/proc-macro-crate-1.2.0/src/lib.rs:137:5
stack backtrace:
   0: rust_begin_unwind
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/panicking.rs:142:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/panicking.rs:181:5
   4: proc_macro_crate::crate_name
             at /Users/djc/.cargo/registry/src/github.com-1ecc6299db9ec823/proc-macro-crate-1.2.0/src/lib.rs:137:5
   5: async_graphql_derive::utils::get_crate_name
             at /Users/djc/.cargo/registry/src/github.com-1ecc6299db9ec823/async-graphql-derive-4.0.6/src/utils.rs:40:26
   6: async_graphql_derive::simple_object::generate
             at /Users/djc/.cargo/registry/src/github.com-1ecc6299db9ec823/async-graphql-derive-4.0.6/src/simple_object.rs:29:22
   7: async_graphql_derive::derive_simple_object
             at /Users/djc/.cargo/registry/src/github.com-1ecc6299db9ec823/async-graphql-derive-4.0.6/src/lib.rs:52:11
   8: core::ops::function::FnOnce::call_once
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/ops/function.rs:248:5
   9: proc_macro::bridge::client::Client<fn(proc_macro::TokenStream) .> proc_macro::TokenStream>::expand1::run::{{closure}}
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/proc_macro/src/bridge/client.rs:424:40
  10: proc_macro::bridge::client::run_client::{{closure}}::{{closure}}
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/proc_macro/src/bridge/client.rs:392:26
  11: proc_macro::bridge::scoped_cell::ScopedCell<T>::set::{{closure}}
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/proc_macro/src/bridge/scoped_cell.rs:79:33
  12: proc_macro::bridge::scoped_cell::ScopedCell<T>::replace
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/proc_macro/src/bridge/scoped_cell.rs:74:9
  13: proc_macro::bridge::scoped_cell::ScopedCell<T>::set
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/proc_macro/src/bridge/scoped_cell.rs:79:9
  14: proc_macro::bridge::client::<impl proc_macro::bridge::Bridge>::enter::{{closure}}
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/proc_macro/src/bridge/client.rs:340:35
  15: std::thread::local::LocalKey<T>::try_with
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/thread/local.rs:445:16
  16: std::thread::local::LocalKey<T>::with
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/thread/local.rs:421:9
  17: proc_macro::bridge::client::<impl proc_macro::bridge::Bridge>::enter
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/proc_macro/src/bridge/client.rs:340:9
  18: proc_macro::bridge::client::run_client::{{closure}}
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/proc_macro/src/bridge/client.rs:385:9
  19: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/core/src/panic/unwind_safe.rs:271:9
  20: std::panicking::try::do_call
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/panicking.rs:492:40
  21: ___rust_try
  22: std::panicking::try
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/panicking.rs:456:19
  23: std::panic::catch_unwind
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/std/src/panic.rs:137:14
  24: proc_macro::bridge::client::run_client
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/proc_macro/src/bridge/client.rs:384:5
  25: proc_macro::bridge::client::Client<fn(proc_macro::TokenStream) .> proc_macro::TokenStream>::expand1::run
             at /rustc/e092d0b6b43f2de967af0887873151bb1c0b18d3/library/proc_macro/src/bridge/client.rs:424:13
  26: proc_macro_srv::abis::abi_1_58::proc_macro::bridge::server::run_server
  27: proc_macro_srv::abis::abi_1_58::Abi::expand
  28: proc_macro_srv::dylib::Expander::expand
  29: core::ops::function::FnOnce::call_once{{vtable.shim}}
```

It looks like #20 is making assumptions it should not be making. Rust-Analyzer issue: https://github.com/rust-lang/rust-analyzer/issues/12969.

cc @jplatte

### PR DESCRIPTION
**Pull request by jplatte**
Fixes #22.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -108,47 +122,64 @@ pub enum FoundCrate {
...
+    })
+}
+
+fn read_cargo_toml(manifest_path: &Path, manifest_ts: SystemTime) -> Result<CacheEntry, Error> {
+    let manifest = open_cargo_toml(manifest_path)?;
+    let crate_names = extract_crate_names(&manifest)?;
+
+    Ok(CacheEntry {
+        manifest_ts,
+        crate_names,
```
Comment (djc) [inline src/lib.rs:179]: Style nit: could directly assign the `crate_names` field instead of binding it separately.

**Review by djc — COMMENTED** [review]
Thanks for fixing this! Here are some suggestions (drive-by review).

File: src/lib.rs [hunk 2]
```diff
@@ -108,47 +122,64 @@ pub enum FoundCrate {
...
         .ok_or_else(|| Error::CrateNotFound {
             crate_name: orig_name.to_owned(),
-            path: cache.manifest_path.clone(),
+            path: manifest_path,
         })?
         .clone())
 }
 
+fn cargo_toml_timestamp(manifest_path: &Path) -> Result<SystemTime, Error> {
+    let timestamp = || fs::metadata(manifest_path)?.modified();
```
[the remark this answers is not in the document]
Comment (jplatte) [reply src/lib.rs:160]: Yes, basically `try`. I guess in this case with just one `?`, `and_then` is a good alternative.

File: src/lib.rs [hunk 3]
```diff
@@ -91,6 +93,18 @@ pub enum FoundCrate {
     Name(String),
 }
 
+// In a rustc invocation, there will only ever be one entry in this map, since every crate is
+// compiled with its own rustc process. However, the same is not (currently) the case for
+// rust-analyzer.
+type Cache = BTreeMap<String, CacheEntry>;
```
[the remark this answers is not in the document]
Comment (jplatte) [reply src/lib.rs:99]: Hashing is entirely useless in the rustc use case, and anything proc-macro related carries the risk of screwing with rust's caches if it uses non-deterministic things like a hash map. I know the non-deterinistic iteration doesn't actually surface here, but I consider it a good rule of thumb to just stay away from `Hash(Map|Set)` in proc-macro crates entirely.

[the remark this answers is not in the document]
Comment (jplatte) [reply src/lib.rs:99]: Re. having a type alias, here it provides a nice place to put the comment about rust-analyzer though, which is unrelated to the comment above the `CACHE` declaration and I thus prefer to be separate.

File: src/lib.rs [hunk 4]
```diff
@@ -108,47 +122,64 @@ pub enum FoundCrate {
...
-            return Err(Error::NotFound(manifest_dir.into()));
-        }
+    // This `Lazy<Mutex<_>>` can just be a `Mutex<_>` starting in Rust 1.63:
+    // https://doc.rust-lang.org/beta/std/sync/struct.Mutex.html#method.new
+    static CACHE: Lazy<Mutex<Cache>> = Lazy::new(Mutex::default);
+    let mut cache = CACHE.lock().unwrap();
 
-        let manifest = open_cargo_toml(&manifest_path)?;
-        let crate_names = extract_crate_names(&manifest)?;
+    let crate_names = match cache.entry(manifest_dir) {
```
[the remark this answers is not in the document]
Comment (jplatte) [reply src/lib.rs:133]: I find the code more readable this way.

File: src/lib.rs (hunk 1 above)
Comment (jplatte) [reply src/lib.rs:179]: Same thing, I thinkk the separate binding makes it easier to read. Would also be shorter if rustfmt would put this struct literal on one line, which it unfortunately doesn't do in the default configuration.

**Review by djc — APPROVED** [review]

**Review by bkchr — APPROVED** [review]
Ty for the fast fix :)

**Pull request merged**